### ISSUE TITLE
[JAWS] Cannot escape out of forms mode in "Both Hands" interaction

### ISSUE THREAD

**Issue by zepumph** [phetsims/ratio-and-proportion#200]
@terracoda mentioned during this morning's description meeting that a learner had trouble with exiting the "Both Hands" interaction using JAWS. She was in a non-standard forms setting where entering/exiting weren't automatic. From my understanding, she was able to press enter to move from the virtual cursor mode into forms mode on the Both Hands application role/div, and use the interaction well. When it came time to leave though, she tried to use the escape key, and it wouldn't work. The only thing that did work was to tab away.

I brought this up with @jessegreenberg today, and we were able to reproduce this behavior in all settings of forms mode control: automatic (default), semi-automatic, and manual.

<details>

```html

<!DOCTYPE html>
<html>
<head>
  <meta charset="UTF-8">
  <title>TEST PAGE</title>
</head>
<body>
<h1>Manual Forms Mode Demo</h1>
<div role="application" tabindex="0" id="focusable">Focus Me!</div>
<p>Events:</p>
<ul id="event-list"></ul>
</body>
<script>
  const focusableElement = document.getElementById( 'focusable' );
  const eventList = document.getElementById( 'event-list' );
  focusableElement.addEventListener( 'keydown', event => {
    const listItem = document.createElement( 'li' );
    listItem.textContent = event.type + ', ' + event.code;
    eventList.appendChild( listItem );
  } );
</script>
</html><!DOCTYPE html>
<html>
<head>
  <meta charset="UTF-8">
  <title>TEST PAGE</title>
</head>
<body>
<h1>Manual Forms Mode Demo</h1>
<div role="application" tabindex="0" id="focusable">Focus Me!</div>
<p>Events:</p>
<ul id="event-list"></ul>
</body>
<script>
  const focusableElement = document.getElementById( 'focusable' );
  const eventList = document.getElementById( 'event-list' );
  focusableElement.addEventListener( 'keydown', event => {
    const listItem = document.createElement( 'li' );
    listItem.textContent = event.type + ', ' + event.code;
    eventList.appendChild( listItem );
  } );
</script>
</html>

```

</details>

In the above snippet, we noticed that the application role got the `esc` key once we had "entered" forms mode on that element. This was the case no matter the forms mode setting.

@jessegreenberg and I feel like this behavior is expected, because the escape key could be part of the interaction, and it would be important to get that event. For example, we rely on getting the escape key event for the `GrabDragInteraction` to "release" the draggable and turn back into a grabbable.

We recommend no further course of action, but want to have @terracoda review and let us know what she thinks should be done from here.

**Comment by terracoda** [comment]
@zepumph, I am not sure I understand. You said, 
>> In the above snippet, we noticed that the application role got the Esc key once we had "entered" forms mode on that element. 

What does that mean? Did you add that? Or have you determined that no action is required?

I agree that the Esc Key can be part of an interaction as you described, going from draggable back to grabbable.

In the case of the Both Hands interaction the hands are "draggable" by default, so how do we make sure the are let-go-able by default as well? If you think about "manual forms mode" nothing is interactable by default, the user always has to do something to enter forms mode, at least that's how the user described it.

In the case of a grab/drag interaction, the user intentionally enters into application role to create the ability to drag, and escapes or tabs away to exit the draggable state.

The question remains, how does one exit an auto-draggable interaction? Is tabbing away the only option? That option was not intuitive to the user, or should we reconsider the auto-draggable interaction?

**Comment by terracoda** [comment]
Or maybe, you meant the Escape key is implemented and is working?

**Comment by terracoda** [comment]
I can't test JAWS, so I cannot actually verify a change it a change was made.
Please enlighten me a little further.

**Comment by zepumph** [comment]
>>In the above snippet, we noticed that the application role got the Esc key once we had "entered" forms mode on that element.

>What does that mean? Did you add that? Or have you determined that no action is required?

I didn't explain well. I'm trying to talk about who "gets" an event coming from the keyboard. I think of the screen reader as a gate keeper for events. It reads each event as it comes, but only certain events get sent through to the webpage. When using the virtual cursor with a screen reader, that screen reader "gets" the arrow key events, and handles them in the screen reader program. It "absorbs" them in a sense, because the arrow key events don't then get passed to the webpage you are using (the virtual-cursor-highlighted `div` element won't get an arrow key event). This applies to the application role because that role tells the screen reader to not absorb any events. Open the gates and don't hold anything back. Once JAWS is in forms mode selected on the application role, none of the key events are "absorbed" by the screen reader, and instead they all get sent through to the webpage. Thus `esc` will never do anything for JAWS, only for the application we have (the boths hands interaction). We could add a listener to the escape key internally, but I don't know a way to trigger JAWS out of forms mode. At least not one that wouldn't potentially be confusing to a user (remove the application role, change the tagName to be something non-interactive, etc). I definitely don't think there is a way for a web page to tell JAWS to get out of forms mode (like from an escape key event listener).

If this doesn't make sense, we should just have a conversation. That will be easier.

**Comment by zepumph** [comment]
>That option was not intuitive to the user, or should we reconsider the auto-draggable interaction?

I'd like to hear if @jessegreenberg has any ideas about a way to have an immediately draggable interaction still be "let-go-able". Feel free to comment, or we can discuss next dev meeting.

**Comment by terracoda** [comment]
@zepumph, thank you for the thoughtful explanation. I do indeed understand now.

This, indeed complicates our ideas about making the hands immediately draggable. The two-step approach we use in other grabbable-draggables provides a simple and direct release operation.

We might have to change the Both Hands interaction into our traditional grab-then-drag interaction.

Currently, if JAWS can't exit forms mode, we seem to be creating a keyboard trap. 

The only idea that comes to my mind is to use the ESC to exit role application, i.e. release the Both Hands, and maybe put focus back on a single hand slider. This would effectively allow the user to press ESC again to exit forms mode. It also allows them to go back and forth between moving both hands and a single hand, though, one can move a single hand in the both hands interaction.

It is still kind of hacky, but two presses of the ESC key might be more intuitive than a Tab away from the entire  interaction with the hands.

If we do have a release...here are some potential context responses:
- Exited both hands mode.
- Both hands released.
- Both hands released, move hands one at a time.
- Both hands released, move {{left/right}} hand. (depending on where focus goes)

Then either a left or right hand focus even will naturally occur: 
- {{current-single-hand-value}}, Left Hand, slider.
- {{current-single-hand-value}}, Right Hand, slider.

I am also very curious if @jessegreenberg has any great ideas.

**Comment by jessegreenberg** [comment]
Hmm, I can't think of a technical way other than what @zepumph suggested (remove the application role, change the tagName to be something non-interactive, etc)

Not sure if it is cause for concern or changes though, it sounds like the user was able to move somewhere else with tab navigation. It doesn't feel as harmful as a keyboard trap to me.

**Comment by zepumph** [comment]
I think before we implement any solution, we should do some research into other application roles online, and see what other behavior is for widgets or components. How do they manage escape?

**Comment by zepumph** [comment]
I had the thought of listening to `escape` key and bluring the interaction. This would send the user to the top of the page though (because focus moves to the body), so we shouldn't do this.

@terracoda had an idea of adding the `application` role dynamically when focusing, and then removing it when you press escape. That may be an appropriate way to handle this.

**Comment by zepumph** [comment]
@terracoda just had another interview with a JAWs user and said that this wasn't an issue. @terracoda can you comment on what forms-mode setting they were using, and what you recommend for this issue?

**Comment by terracoda** [comment]
In latest interview where I did not notice any issues around "getting stuck" on the Both hands interview, the participant said they use JAWS with forms mode off. I think this means in "manual" mode.

In the interview, where the issue occurred the forms mode setting was described as "manual".

So, in the latest version of the sim this issue did not come up, and unfortunately, I did not specifically ask about how the Both Hands interaction was working for the participant. There were no obvious "getting stuck points" while interacting with Both Hands.

**Comment by terracoda** [comment]
There was a moment in the latest interview where the participant was on the Both Hands interaction and the interaction seemed activated, and the participant was not able to access the help text for the interaction. I'll review the section in the interview to see if I can make sense of it and to see if it is related to this issue. 

In any case, the situation was not like one that caused me to open this issue.

**Comment by terracoda** [comment]
Ok, I reviewed the video. 
The issue is once focus is on the Both Hands, the down arrow key only works to move the right hand down. The only way out of the interaction is with the Tab key.

It might be helpful for this sim if the instructions (the help text) actually come before the interactions. I this would be beneficial for both the sliders and the Both Hands interactions.

Can we re-arrange things like this?

H2: Play Area
P: Move hands up and down individually to find the Challenge Ratio in different places.
Label: Left Hand
hands, very close to challenge ratio                  

Label: Right Hand
hands, very close to challenge ratio                                                        

P: **With Both Hands activated,** use W and S keys to move left hand, and Up and Down Arrow keys to move right hand. Move both hands at the same time to find and maintain the Challenge Ratio.

DIV(with custom stuff):Both Hands
P: hands very close to each other _(this text should be the object value of both hands. I'll make a new issue for that.)_

**Comment by terracoda** [comment]
@zepumph in 4176e265 I adjusted the help text for the Both Hands interaction, and changed "Both hands" to "Both Hands".

Would it be a lot of work to change the PDOM order?

**Comment by terracoda** [comment]
Assigning to you @zepumph to rearrange things.

**Comment by terracoda** [comment]
In https://github.com/phetsims/ratio-and-proportion/issues/262#issuecomment-748277232 I decided we needed a dynamic description of the hands, which is somewhat related to what I was proposing above in https://github.com/phetsims/ratio-and-proportion/issues/200#issuecomment-742620175

I am going to close this issue as adding and open a new one that describes adding a dynamic description and re-arrangement of Play Area content.
